ma_gym/envs/traffic_junction/traffic_junction.py

```diff
@@ -73,13 +116,96 @@ def __init__(self, grid_shape=(14, 14), n_agents=10, n_max=10, arrive_prob=0.5,
         if self.full_observable:
             self._obs_high = np.tile(self._obs_high, self.n_agents)
             self._obs_low = np.tile(self._obs_low, self.n_agents)
-        self.observation_space = MultiAgentObservationSpace([spaces.Box(self._obs_low, self._obs_high) for _ in range(self.n_agents)])
-        self.seed()
+        self.observation_space = MultiAgentObservationSpace([spaces.Box(self._obs_low, self._obs_high)
+                                                             for _ in range(self.n_agents)])
 
     def action_space_sample(self):
         return [agent_action_space.sample() for agent_action_space in self.action_space]
 
+    def __init_full_obs(self):
+        """
+        Initiates environment: inserts up to |entry_gates| cars. once the entry gates are filled, the remaining agents
+        stay initialized outside the road waiting to enter
+        """
+        self._full_obs = self.__create_grid()
+
+        for agent_i in range(self.n_agents):
+            while True:
```

Comment:
In general, try to avoid infinite while loops.

Superficially, I think, this code could be re-written to avoid a while loop.

Steps: instead of using the random.choice, just use `random.shuffle` and get a  shuffled list of  route_vectors and simply assign these route vectors to the first 4 (or max. no, of entry gates)  agents.  Thereafter, assign "not yet on the road" in a new "for" loop.
- Also, this way, you wouldn't have to check for "is_cell_vacant"